framework/include/times/Times.h

```diff
@@ -0,0 +1,76 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+// Moose includes
+#include "GeneralReporter.h"
+
+/**
+ * Times objects are under the hood Reporters, but limited to a vector of Real
+ */
+class Times : public GeneralReporter
+{
+public:
+  static InputParameters validParams();
+  Times(const InputParameters & parameters);
+  virtual ~Times() = default;
+
+  /// Getter for the full times vector
+  const std::vector<Real> & getTimes() const;
+
+  /// Getter for a set of the full times
+  const std::set<Real> getUniqueTimes() const
+  {
+    return std::set<Real>(_times.begin(), _times.end());
+  };
+
+  /// Get the current time
+  Real getCurrentTime() const { return _fe_problem.time(); }
+
+  /// Getter for a single time at a known index
+  Real getTimeAtIndex(unsigned int index) const;
+
+  /// Find the previous time in the times vector for a given time
+  /// If current_time is also in the times vector within numerical precision, will return the previous value
+  Real getPreviousTime(const Real current_time) const;
+
+  /// Find the next time in the times vector for a given time
+  /// If current_time is also in the times vector within numerical precision, will return the next value
+  Real getNextTime(const Real current_time) const;
+
+protected:
+  /// In charge of computing / loading the times, unless all that could be done there is done
+  /// in the constructor
+  virtual void initialize() override = 0;
+
+  /// By default, we wont execute often but "executing" will mean loading the times
+  virtual void execute() override { initialize(); }
+
+  /// In charge of reduction across all ranks
+  virtual void finalize() override;
+
+  /// By default, Times will not be modified very regularly
+  virtual void timestepSetup() override {}
+  virtual void residualSetup() override {}
+  virtual void jacobianSetup() override {}
+
+  /// Clear the times vector
+  void clearTimes();
+
+  /// The vector holding the times
+  std::vector<Real> & _times;
+
+  /// Whether generation of times is distributed or not (and therefore needs a broadcast)
+  bool _need_broadcast;
+  /// Whether times should be sorted, because they come from different sources for example
+  bool _need_sort;
+  /// Whether duplicate times should be removed
+  bool _need_unique;
```

Comment:
would be unnecessary if we just use a set

Reply:
Even with `std::set`, I think @GiudGiud is onto something with the "floating point comparisons issues" comment in another thread ... but if we're expecting to be computing times with FP roundoff error, then do we want "uniqueness" up to some specified tolerance, not just a bool?  A `set` will give unique times with exact comparisons, but if we have a `t=0.2` and another `t=0.02*10` then we'll need some `epsilon > 3e-17` to tell the system that we really think those two unique-in-FP numbers should be considered equivalent.

Reply:
It's not really about uniqueness to me though, it's about whether it's sorted

Reply:
But the concern you raise is valid

Reply:
I ll put up some patch for tolerance. 

I dont think I want them sorted by default.

Reply:
why not?

Reply:
see reasons above plus it's really easy to think of a use case where we would want more than one time. Let me give examples here and I can come up with more if still not convinced 

- if we were to keep track of when a simulation failed. The first time it failed a step could be at t=100, we try again with a cut step at t=50. If we fail again then pass at 25, we loose track of how it unfolded. If we kept times unique, we can even loose track of how many times we failed a certain time

- we have multiple subapps contributing times to a Times object that is keeping track of an event. Subapp 1 says t=25, subapp 2 says t=20. Then if we sort, we loose track that subapp 2 had the lower number

- we run a transient direct then an adjoint calculation (with the reverse time). If we sort, we loose track that the times in the adjoint simulation came after, they will all be mixed together

Reply:
alright, those are great examples 😄

Reply:
A perfect use case for a `std::vector`: a sequence!